flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/BuiltInSqlFunction.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions.sql;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.functions.BuiltInFunctionDefinition;
+import org.apache.flink.table.planner.functions.bridging.BridgingSqlFunction;
+
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.type.SqlOperandTypeChecker;
+import org.apache.calcite.sql.type.SqlOperandTypeInference;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.validate.SqlMonotonicity;
+
+import javax.annotation.Nullable;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import static org.apache.flink.table.functions.BuiltInFunctionDefinition.DEFAULT_VERSION;
+import static org.apache.flink.table.functions.BuiltInFunctionDefinition.qualifyFunctionName;
+import static org.apache.flink.table.functions.BuiltInFunctionDefinition.validateFunction;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * SQL version of {@link BuiltInFunctionDefinition} in cases where {@link BridgingSqlFunction} does
+ * not apply. This is the case when the operator has a special parsing syntax or uses other
+ * Calcite-specific features that are not exposed via {@link BuiltInFunctionDefinition} yet.
+ *
+ * <p>Note: Try to keep usages of this class to a minimum and use Flink's {@link
+ * BuiltInFunctionDefinition} stack instead.
+ *
+ * <p>For simple functions, use the provided builder. Otherwise, this class can also be extended.
+ */
+@Internal
+public class BuiltInSqlFunction extends SqlFunction implements BuiltInSqlOperator {
```

Comment:
Can you move the changes related to this class (and `BuiltInSqlOperator`) in a separate commit, as they seem more of a refactoring hotfix instead?

Reply:
This change is part of the internal/versioning story. However, I will split the refactorings into a separate commit.